frame/collective/src/lib.rs

```diff
@@ -182,72 +306,200 @@ decl_module! {
-			Self::deposit_event(RawEvent::MemberExecuted(proposal_hash, ok));
+			let result = proposal.dispatch(RawOrigin::Member(who).into());
+			Self::deposit_event(RawEvent::MemberExecuted(proposal_hash, result.is_ok()));
+
+			Ok(get_result_weight(result).map(|w| weight_for::execute(
+				T::DbWeight::get(),
+				members.len() as Weight,
+				w
+			)).into())
 		}
 
+		/// Add a new proposal to either be voted on or executed directly.
+		///
+		/// Requires the sender to be member.
+		///
+		/// `threshold` determines whether `proposal` is executed directly (`threshold < 2`)
+		/// or put up for voting.
+		///
 		/// # <weight>
-		/// - Bounded storage reads and writes.
-		/// - Argument `threshold` has bearing on weight.
+		/// ## Weight
+		/// - `O(B + M + P1)` or `O(B + M + P2)` where:
+		///   - `B` is `proposal` size in bytes (length-fee-bounded)
+		///   - `M` is members-count (code- and governance-bounded)
+		///   - branching is influenced by `threshold` where:
+		///     - `P1` is proposal execution complexity (`threshold < 2`)
+		///     - `P2` is proposals-count (code-bounded) (`threshold >= 2`)
+		/// - DB:
+		///   - 1 storage read `is_member` (codec `O(M)`)
+		///   - 1 storage read `ProposalOf::contains_key` (codec `O(1)`)
+		///   - DB accesses influenced by `threshold`:
+		///     - EITHER storage accesses done by `proposal` (`threshold < 2`)
+		///     - OR proposal insertion (`threshold <= 2`)
+		///       - 1 storage mutation `Proposals` (codec `O(P2)`)
+		///       - 1 storage mutation `ProposalCount` (codec `O(1)`)
+		///       - 1 storage write `ProposalOf` (codec `O(B)`)
+		///       - 1 storage write `Voting` (codec `O(M)`)
+		///   - 1 event
+		/// - Benchmarks:
+		///   - execute: 28.12 + M * 0.218 + B * 0.003 µs (min squares analysis) (`threshold < 2`)
+		///   - propose: 49.75 + M * 0.105 + P2 0.502 + B * 0.006 µs (min squares analysis) (`threshold >= 2`)
 		/// # </weight>
-		#[weight = (5_000_000_000, DispatchClass::Operational)]
-		fn propose(origin, #[compact] threshold: MemberCount, proposal: Box<<T as Trait<I>>::Proposal>) {
+		#[weight = (
+			if *threshold < 2 {
+				weight_for::propose_execute(
+					T::DbWeight::get(),
+					T::MaxMembers::get(), // M
+					proposal.get_dispatch_info().weight, // P1
+				)
+			} else {
+				weight_for::propose_proposed(
+					T::DbWeight::get(),
+					T::MaxMembers::get(), // M
+					T::MaxProposals::get() // P2
+				)
+			},
+			DispatchClass::Operational
+		)]
+		fn propose(
+			origin, #[compact] threshold: MemberCount, proposal: Box<<T as Trait<I>>::Proposal>
+		) -> DispatchResultWithPostInfo {
 			let who = ensure_signed(origin)?;
-			ensure!(Self::is_member(&who), Error::<T, I>::NotMember);
+			let members = Self::members();
+			ensure!(members.contains(&who), Error::<T, I>::NotMember);
 
 			let proposal_hash = T::Hashing::hash_of(&proposal);
-
 			ensure!(!<ProposalOf<T, I>>::contains_key(proposal_hash), Error::<T, I>::DuplicateProposal);
```

Comment:
@shawntabrizi The hash is used to check for duplicate proposals. Not sure how essential that is.